### PR TITLE
Implement np.linalg.pinv

### PR DESCRIPTION
**Pull request by TuanNguyen27**
Attempting #1638

### CODE REVIEW

File: tests/linalg_test.py [hunk 1]
```diff
@@ -559,6 +559,27 @@ def args_maker():
...
+      for dtype in float_types
+      for rng in [jtu.rand_default()]))
+    def testPinv(self, shape, dtype, rng):
+      _skip_if_unsupported_type(dtype)
+      if jtu.device_under_test() == "gpu" and shape == (200, 200):
+        raise unittest.SkipTest("Test is flaky on GPU")
+
+      def args_maker():
+        a = rng(shape, dtype)
+        onp.linalg.pinv(a)
```
Comment (shoyer) [inline tests/linalg_test.py]: This line looks like a no-op, remove?

File: jax/numpy/linalg.py [hunk 2]
```diff
@@ -136,6 +136,21 @@ def eigvalsh(a, UPLO='L'):
...
+  a = np.conj(a)
+  rcond = np.asarray(rcond)
+  u, s, v = svd(a, full_matrices=False)
+  # Singular values less than or equal to ``rcond * largest_singular_value`` 
+  # are set to zero.
+  cutoff = rcond[..., np.newaxis] * np.amax(s, axis=-1, keepdims=True)
+  large = s > cutoff
+  s = np.divide(1, s)
+  s = np.where(large, s, 0)
+  res = np.matmul(np.transpose(v), np.multiply(s[..., np.newaxis], np.transpose(u)))
```
Comment (shoyer) [inline jax/numpy/linalg.py]: Note that `transpose` function that NumPy uses here is not *quite* the same as `np.transpose` -- it only flips the last two axes. Is this version still correct?

File: tests/linalg_test.py [hunk 3]
```diff
@@ -559,6 +559,27 @@ def args_maker():
                             check_dtypes=True, tol=1e-3)
     self._CompileAndCheck(np.linalg.inv, args_maker, check_dtypes=True)
 
+    @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}".format(jtu.format_shape_dtype_string(shape, dtype)),
+       "shape": shape, "dtype": dtype, "rng": rng}
+      for shape in [(1, 1), (4, 4), (2, 5, 5), (200, 200), (5, 5, 5)]
+      for dtype in float_types
```
Comment (shoyer) [inline tests/linalg_test.py]: Please add tests covering:
- Complex dtypes. I see at least one place where your code is missing `.conjugate()` vs NumPy's implementation, so I guess this would currently fail.
- More array shapes, at least including a wide matrix and a tall matrix.

**Review by shoyer — COMMENTED** [review]
Thanks for putting this together! I have a concerns, noted before.

I assume you based this off of the NumPy implementation? That's fine, just good to note for reference :)
https://github.com/numpy/numpy/blob/v1.17.0/numpy/linalg/linalg.py#L1890-L1979

File: tests/linalg_test.py (hunk 1 above)
Comment (TuanNguyen27) [reply tests/linalg_test.py]: you're right, will do !

File: jax/numpy/linalg.py (hunk 2 above)
Comment (TuanNguyen27) [reply jax/numpy/linalg.py]: Good catch! in this case `np.swapaxes` would be more appropriate?

**Comment by TuanNguyen27** [comment]
@shoyer thanks a lot for reviewing the PR ! Will address your comments soon.

File: jax/numpy/linalg.py [hunk 4]
```diff
@@ -136,6 +136,24 @@ def eigvalsh(a, UPLO='L'):
   return w
 
 
+@_wraps(onp.linalg.pinv)
+def pinv(a, rcond=1e-15):
```
[the remark this answers is not in the document]
Comment (TuanNguyen27) [reply jax/numpy/linalg.py]: good point, how should i pick a better threshold ?

[the remark this answers is not in the document]
Comment (shoyer) [reply jax/numpy/linalg.py]: I think we could justify copying the default behavior from tensorflow:
"This function is analogous to numpy.linalg.pinv. It differs only in default value of rcond. In numpy.linalg.pinv, the default rcond is 1e-15. Here the default is `10. * max(num_rows, num_cols) * np.finfo(dtype).eps`."
https://www.tensorflow.org/probability/api_docs/python/tfp/math/pinv

[the remark this answers is not in the document]
Comment (TuanNguyen27) [reply jax/numpy/linalg.py]: thanks for the pointer!

[the remark this answers is not in the document]
Comment (TuanNguyen27) [reply jax/numpy/linalg.py]: @shoyer I've added the same default as `tfp`, but I'm not confident in how I handle mal-formed / 1-dim `a`. Could you take a look? Thanks! :)

**Comment by duckworthd** [comment]
Thank you for this implementation, @TuanNguyen27!

May I also request a test for the ill-conditioned case? My own attempts at implementing `pinv()` fall apart when the condition number exceeds 10^5 on float32 precision...

**Comment by TuanNguyen27** [comment]
> Thank you for this implementation, @TuanNguyen27!
> 
> May I also request a test for the ill-conditioned case? My own attempts at implementing `pinv()` fall apart when the condition number exceeds 10^5 on float32 precision...

Sure! Let me add a test case where the size of matrix matches your test case :)

**Comment by TuanNguyen27** [comment]
@duckworthd Tests are passing with matrices of size 7e5 and 7e6 ! I think my test covers both float32 and float64.

**Comment by duckworthd** [comment]
Fantastic, thank you!

File: jax/numpy/linalg.py [hunk 5]
```diff
@@ -136,6 +136,32 @@ def eigvalsh(a, UPLO='L'):
...
+def pinv(a, rcond=None):
+  # ported from https://github.com/numpy/numpy/blob/v1.17.0/numpy/linalg/linalg.py#L1890-L1979
+  a = np.conj(a)
+  if rcond is None:
+      if np.ndim(a) < 2:
+        num_rows = num_cols = np.shape(a)[-1]
+      else:
+        num_rows = np.shape(a)[-2]
+        num_cols = np.shape(a)[-1]
+      max_rows_cols = np.max(num_rows, num_cols)
```
Comment (shoyer) [inline jax/numpy/linalg.py]: I  would suggest just writing `max_row_cols = max(a.shape[-2:])` (the Python builtin max, not numpy's max). That will work on arrays of any shape.

File: jax/numpy/linalg.py [hunk 6]
```diff
@@ -136,6 +136,32 @@ def eigvalsh(a, UPLO='L'):
...
+      rcond = 10. * max_rows_cols * np.finfo(a.dtype)
```
Comment (shoyer) [inline jax/numpy/linalg.py]: You need pull off the appropriate attribute from `finfo`, e.g.,
```suggestion
      rcond = 10. * max_rows_cols * np.finfo(a.dtype).eps
```

I'm actually rather puzzled by what is going on here -- how could your unit tests be possible passing? Are you sure they are actually being run?

File: jax/numpy/linalg.py [hunk 7]
```diff
@@ -136,6 +136,32 @@ def eigvalsh(a, UPLO='L'):
   return w
 
 
+@_wraps(onp.linalg.pinv)
+def pinv(a, rcond=None):
+  # ported from https://github.com/numpy/numpy/blob/v1.17.0/numpy/linalg/linalg.py#L1890-L1979
+  a = np.conj(a)
+  if rcond is None:
+      if np.ndim(a) < 2:
```
Comment (shoyer) [inline jax/numpy/linalg.py]: I don't think pinv is well defined on 1d arrays, so you should not need this logic.

File: jax/numpy/linalg.py (hunk 6 above)
Comment (TuanNguyen27) [reply jax/numpy/linalg.py]: Ah, i think i messed up the indentation of my test and they weren't being run. They should be running now, but failing on big matrix cases. :(

File: tests/linalg_test.py [hunk 8]
```diff
@@ -577,6 +577,22 @@ def args_maker():
                             check_dtypes=True, tol=1e-3)
     self._CompileAndCheck(np.linalg.inv, args_maker, check_dtypes=True)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}".format(jtu.format_shape_dtype_string(shape, dtype)),
+       "shape": shape, "dtype": dtype, "rng_factory": rng_factory}
+      for shape in [(1, 1), (4, 4), (2, 70, 7), (int(7e5), 70), (70, int(7e6)), (70, 7, 2)]
```
Comment (shoyer) [inline tests/linalg_test.py]: Some of these arrays are too big for testing purposes. 7e5 x 70 elements is about 400 MB in complex128 and 70 x 7e5 elements is 10x larger! This is causing the tests on Travis-CI to fail.

We definitely don't want to require this much free memory to test JAX. I would suggest sticking to a maximum array size of no more than 1e6 or so total elements. But  even that is likely bigger than  we really need.

Given  that  this test does not implement any novel numerical methods, it doesn't need to test numerics  exhaustively. That's what XLA's own unit tests are for.

File: tests/linalg_test.py [hunk 9]
```diff
@@ -577,6 +577,22 @@ def args_maker():
                             check_dtypes=True, tol=1e-3)
     self._CompileAndCheck(np.linalg.inv, args_maker, check_dtypes=True)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}".format(jtu.format_shape_dtype_string(shape, dtype)),
+       "shape": shape, "dtype": dtype, "rng_factory": rng_factory}
+      for shape in [(1, 1), (4, 4), (2, 70, 7), (2e4, 7), (7, 1e5), (70, 7, 2)]
```
Comment (shoyer) [inline tests/linalg_test.py]: Floats aren't valid in NumPy shape parameters.

Try:
```suggestion
      for shape in [(1, 1), (4, 4), (2, 70, 7), (int(2e4), 7), (7, int(1e5)), (70, 7, 2)]
```
or:
```suggestion
      for shape in [(1, 1), (4, 4), (2, 70, 7), (2000, 7), (7, 10000), (70, 7, 2)]
```

Comment (TuanNguyen27) [reply tests/linalg_test.py]: thanks for the detailed feedbacks, @shoyer !

Comment (TuanNguyen27) [reply tests/linalg_test.py]: Tests are failing with a dtype mismatch (results returning `float64` instead of `float32`), but only for `(70,7,2)` case...and only for certain versions of Python. Do you have an idea what could go wrong here ?

Comment (TuanNguyen27) [reply tests/linalg_test.py]: Ah, I suspect this is related to #768. Would it be okay if I default `onp.linalg.pinv` to return `float64` instead? -- This might instead break things when jax.numpy actually returns `float32`. Let me know what you think!

Comment (mattjj) [reply tests/linalg_test.py]: (@Shoyer @TuanNguyen27 Thanks for your incredible work on this so far, and attention to detail!)

Re: dtype issues and #768, those are tricky to get right, but they should always be solvable by explicitly "casting" Python literals to specific dtypes, using things like `np.float32(10)` in place of just the literal `10`. Maybe the literals in the code need to be cast that way to avoid NumPy automatic promotion?

Comment (TuanNguyen27) [reply tests/linalg_test.py]: Both `np.float32(10)` and `rcond = np.asarray(rcond, dtype=a.dtype)` didn't work, but casting the final result `lax.convert_element_type(res, a.dtype)` seems to do the trick. Is this okay ?

Comment (TuanNguyen27) [reply tests/linalg_test.py]: @shoyer pinging you again in case this is good to merge !

**Comment by dbieber** [comment]
Thanks for all your work on this!
I notice there's been no activity for a few weeks though. I'm also interested in using pinv, so just checking in to see if you still plan on getting this PR checked in.

**Review by shoyer — COMMENTED** [review]
oops, I wrote these comments last week and forgot to send them!

**Comment by shoyer** [comment]
I think this is basically ready, just needs a couple of comment/doc fixes

**Comment by shoyer** [comment]
Could also submit this and add those later :)

**Comment by mattjj** [comment]
@shoyer Since it's been a while on our end, WDYT about landing and iterating? Do you want to take on the remaining todos  (from your comments), or should I? (I don't mind doing it, if it means merging this excellent feature sooner!)

**Comment by shoyer** [comment]
@mattjj sounds good to me -- merged!
